### PR TITLE
Use the Ky request library in a web browser context instead of Got

### PR DESCRIPTION
**Pull request by jlipps**
## Proposed changes

Some time ago, WebdriverIO standardized on Got as the request library. Unfortunately this broke downstream projects like Web2Driver which need to be able to bundle the webdriver package for the web. This PR adds in support for the Ky browser-based request library, and selects it in a non-Node.js environment.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [X] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [X] I have added proper type definitions for new commands (if appropriate)

## Further comments

There are currently no tests for the browser-based client, since I think that would be a pretty big update to the whole test environment. I'm open to hearing how to approach testing (beyond using it in my own project).

### CODE REVIEW

File: packages/webdriver/tests/request.test.ts [hunk 1]
```diff
@@ -314,6 +314,7 @@ describe('webdriver request', () => {
             const req = new WebDriverRequest('POST', path, {})
             req.emit = jest.fn()
 
+            console.log(req.defaultOptions)
```
Comment (christian-bromann) [inline packages/webdriver/tests/request.test.ts:317]:
```suggestion
```

**Review by christian-bromann — CHANGES_REQUESTED** [review]
Thanks for pushing this forward 🙏 

Overall looks good to me, some suggestions:

- can we also have a unit test for `request-browser.ts`?
- can we change the folder structure to the following:

```
/ src
    |- command.ts
    |- /request
        |- index.ts (current request.ts)
        |- node.ts (current request-node.ts)
        |- factory.ts (current request-factory.ts)
        |- browser.ts (current request-browser.ts)
   |- ...
```

**Comment by jlipps** [comment]
Sounds good, I'm working on these changes. I have to mock Ky completely since we don't even want to import it from Node (it's exported in a way that's designed to be bundled)

**Comment by jlipps** [comment]
ok @christian-bromann I added tests to mirror the tests for the node version of the requests, and I've also tested in Web2Driver (that e2e suite passed). There seem to be some CI issues here with conflicting versions of types that I'm not sure how to resolve though.

**Comment by christian-bromann** [comment]
The code looks good. Is the filename `brower.ts` intentional or a typo? I can take a look at the typing issue.

**Comment by christian-bromann** [comment]
@jlipps can you revert your changes in `package-lock.json`? It should resolve the typing problems.

**Comment by jlipps** [comment]
> The code looks good. Is the filename brower.ts intentional or a typo? I can take a look at the typing issue.

Hmm it was an extra file somehow, obviously a typo. I removed it.

> can you revert your changes in package-lock.json? It should resolve the typing problems.

OK, let's see how that works!

**Comment by jlipps** [comment]
@christian-bromann even without any package-lock changes there seem to be some typing conflicts on windows and some versions of node on linux. but on other platforms it looks OK. that's odd, right?

**Comment by christian-bromann** [comment]
Rerunning, just saw it also happening in an other PR.

**Comment by jlipps** [comment]
rerunning doesn't seem to have helped. 🤔

**Comment by christian-bromann** [comment]
@jlipps is this something you can reproduce locally, via:

```sh
$ rm -r node_modules package-lock.json
$ npm install
$ npm run setup-full
# if this fails you can only compile via
$ npm run build
```

?

**Comment by jlipps** [comment]
@christian-bromann all of those commands completed successfully (though of course they resulted in a package-lock.json which differs from the one on `main`). would you like me to run any other commands, or push the lockfile changes?

**Comment by jlipps** [comment]
Any more thoughts on this @christian-bromann? If you pull the branch and run tests, does it work? Want me to push lockfile updates? Or run a lockfile update on `main` so that I can rebase on it?

**Comment by christian-bromann** [comment]
closing in favor of https://github.com/webdriverio/webdriverio/pull/6999

**Pull request closed without merging**